dir.go

```diff
@@ -122,8 +122,8 @@ func (d *Directory) localUpdate(c child) (*dag.ProtoNode, error) {
 		return nil, err
 	}
 
+	// A copy is needed here since nd is referenced by dagService.
```

Comment:
Could you elaborate? I understand the `dagService` takes the node as (read-only) input and process it but as far as I know this happens sequentially: once we reach this line already `dagService` added the node and doesn't hold to the reference anymore.

Reply:
The `dagService`. [adds the block])https://github.com/ipfs/go-merkledag/blob/master/merkledag.go#L51) to the blockstore,  The blockservice [puts](https://github.com/ipfs/go-blockservice/blob/master/blockservice.go#L146) the block into the blockstore.  The blocksstore [puts](https://github.com/ipfs/go-ipfs-blockstore/blob/master/blockstore.go#L169) the raw block data into the datastore.  This [raw data](https://github.com/ipfs/go-block-format/blob/master/blocks.go#L64) is a byte slice from the block, not a copy, and the memory it references leads to a race condition.  I am not sure where that memory gets overwritten, but running tests with the race detector will flag the race condition.

Reply:
I agree with your assessment of the block data model but I still fail to see where is the potential race.

By this point in the code `dagService.Add()` has finished and the component that has access to this byte slice (`BlockService`) has already stored it in the `Datastore`. These components AFAIK do not keep a reference to the slice; in a way _they_ are the ones doing the copying (actually dumping the bytes in the store medium) of the slice, so why do we need to copy as well? Could you point me to the source line where you think the access might cause the race please?

Reply:
It is possible the race condition is only induced by the test code.  I will take more time to see if this is possible outside of the test.

Reply:
`localUpdate` here is an internal function. We are not exposing the returned internal structure in the API. As its documentation describes, `localUpdate` is just an extraction from the bigger (also internal) `updateChildEntry`.